### PR TITLE
develop <- PR-600

### PR DESCRIPTION
**Pull request by nico-isento**

### CODE REVIEW

File: ros_packages/voice_assistant/audiofiles/text_to_speech/2lkso9w6
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/audiofiles/text_to_speech/2lkso9w6:1]: they don't serve a purpose and were commited on accident. They are now removed

File: ros_packages/voice_assistant/pib_voice/pib_voice/voice.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    except sr.RequestError as e:
+        print('Request error from Google Speech Recognition')
+    return data
+
+
+
+def play_audio_from_text(text: str) -> None:
+
+    pya = pyaudio.PyAudio()
+    stream = pya.open(format=pya.get_format_from_width(width=2), channels=1, rate=16000, output=True)
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/pib_voice/pib_voice/voice.py:69]: pyaudio-names were unified

File: ros_packages/voice_assistant/pib_voice/pib_voice/voice.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    
+    stream.write(response['AudioStream'].read())
+
+    stream.stop_stream()
+    stream.close()
+    pya.terminate()
+
+
+
+def play_audio(file_path: str) -> None:
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/pib_voice/pib_voice/voice.py]: files was renamed

File: ros_packages/voice_assistant/pib_voice/pib_voice/voice.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                "content": input_text,
+            },
+        ]
+    )
+
+    return response.choices[0].message.content
+
+
+
+def text_to_speech(text_input: str, gender: str, output_file_dir: str) -> str:
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/pib_voice/pib_voice/voice.py]: no, it became redundant with the introduction of aws-polly. It was removed

File: ros_packages/voice_assistant/voice_assistant/assistant.py [hunk 4]
```diff
@@ -150,165 +146,101 @@ def get_personality_from_chat_id(self, chat_id: str) -> Personality | None:
...
-                            self.state_to_main.send((personality, request_state.chat_id))
-                            self.state = VoiceAssistantState(turned_on=True, chat_id=request_state.chat_id)
-                        response.successful = personality is not None
-                    else:
-                        self.state_to_main.send(None)
-                        response.successful = True
-                        self.state = VoiceAssistantState(turned_on=False, chat_id=request_state.chat_id)
+                    self.tts_clear_client.call(TextToSpeechClear.Request())
+                    with self.ros_to_worker_lock: 
+                        self.worker_changed = True # TODO: worker_changed -> worker_stale
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/voice_assistant/assistant.py]: TODO was removed

File: ros_packages/voice_assistant/voice_assistant/text_to_speech.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        while True:
+            with self.ros_to_worker_lock:
+                if self.worker_connection_stale: 
+                    return response
+                if self.ros_to_worker.poll():
+                    self.ros_to_worker.recv()
+                    return response
+            time.sleep(0.1)
+
+    def clear(self, request: TextToSpeechClear.Request, response: TextToSpeechClear.Response) -> TextToSpeechClear.Response:
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/voice_assistant/text_to_speech.py:84]: no, but ros requires this type of signature for a service-callback-function

File: ros_packages/voice_assistant/voice_assistant/text_to_speech.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+        
+
+def worker_target(worker_to_ros: Connection) -> None:
+
+    while True:
+
+        # receive next text from ros and play back the text as audio
+        request: TextRequest = worker_to_ros.recv()
+        play_audio_from_text(request.text, request.voice)
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/voice_assistant/text_to_speech.py:141]: function accepted the second parameter at some point, but it seems like the change was not commited. Second param was added again

File: ros_packages/voice_assistant/pib_voice/pib_voice/voice.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+from typing import Any
+import pyaudio
+import wave
+import os
+import speech_recognition as sr
+from speech_recognition import WaitTimeoutError
+from openai import OpenAI
+import boto3
+
+
```
[the remark this answers is not in the document]
Comment (nico-isento) [reply ros_packages/voice_assistant/pib_voice/pib_voice/voice.py:14]: it is in order to separate the ros-logic from the logic for communicating with external apis, playing back audio, etc.

**Review by nmsger-isento — APPROVED** [review]

**Pull request merged**